### PR TITLE
Support more than one authentication session

### PR DESCRIPTION
**Pull request by vinokurig**
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

### What does this PR do?
* Handle multiple GitHub authentication sessions.
* Add back the vscode built-in github plugin to be able to publish `https` repositories. 

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/18227

### How to test this PR?

1. Setup [GitHub Identity provider](https://www.eclipse.org/che/docs/che-7/end-user-guide/configuring-github-oauth/)
2. Start a workspace with GitHub Pull Requests plugin and a project cloned from GitHub by `https` url.
You can use this devfile to start a workspace:
```
apiVersion: 1.0.0
metadata:
  name: wksp-custom-dwdiu
projects:
  - name: che-theia
    source:
      location: 'https://github.com/eclipse/che-theia.git'
      type: git
      branch: master
components:
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/eclipse/che-plugin-registry/master/v3/plugins/ms-vscode/vscode-github-pullrequest/0.20.0/meta.yaml'
  - type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-906/simple/che-theia-editor.yaml'
```
3. Sign-in to the GitHub Pull Requests plugin.
4. Checkout to a new branch and commit some changes.
5. Open a new Pull Request, see https://code.visualstudio.com/docs/editor/github#_creating-pull-requests

The GitHub Pull Requests plugin publishes the branch to origin, so an additional GitHub authentication session is requested.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

File: plugins/github-auth-plugin/src/github-auth-plugin.ts [hunk 1]
```diff
@@ -10,45 +10,49 @@
...
-                    return [];
-                }
-            },
+            getSessions: async () => sessions,
             login: async (scopes: string[]) => {
                 const githubUser = await che.github.getUser();
-                session = {
-                    id: 'github-session',
+                const session = {
+                    id: v4(),
```
Comment (benoitf) [inline plugins/github-auth-plugin/src/github-auth-plugin.ts:27]: do we need uuid or just a counter ?

**Comment by azatsarynnyy** [comment]
image

**Review by azatsarynnyy — COMMENTED** [review]
Maybe, I'm doing something wrong, but here's how it looks on my side:

Peek 2020-11-05 22-34

**Comment by vinokurig** [comment]
@azatsarynnyy 
Looks like you are using an old `che-plugin-registry` image, so the GitHub PR plugin is not updated there.
Please set the version of the `che-plugin-registry` image to `nightly`:
1. Run `kubectl edit checluster eclipse-che -n che`
2. Press `a` key and add `pluginRegistryImage: quay.io/eclipse/che-plugin-registry:nightly`
3. Press `Esc` key and enter `wq` to apply changes

**Comment by vinokurig** [comment]
@benoitf 
> do we need uuid or just a counter ?

uuid looks simpler for me.

**Comment by benoitf** [comment]
@vinokurig I think the devfile you're suggesting, you should use `reference:` and link to the master version of plug-in registry meta.yaml so devfile, host it on a gist, and provide the link directly so it can be tested with che.openshift.io without any user action

**Comment by vinokurig** [comment]
@benoitf 
> I think the devfile you're suggesting, you should use reference: and link to the master version of plug-in registry meta.yaml so devfile, host it on a gist, and provide the link directly so it can be tested with che.openshift.io without any user action

Yes you are right, I've edited the description.

**Comment by vinokurig** [comment]
@azatsarynnyy Please use the edited devfile from the description.

**Comment by azatsarynnyy** [comment]
> @azatsarynnyy Please use the edited devfile from the description.

Thanks @vinokurig ! Checking...

**Comment by azatsarynnyy** [comment]
@vinokurig should the test Devfile also include Che Theia Editor from this PR? 
```yaml
apiVersion: 1.0.0
metadata:
  name: wksp-custom-dwdiu
projects:
  - name: che-theia
    source:
      location: 'https://github.com/eclipse/che-theia.git'
      type: git
components:
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/eclipse/che-plugin-registry/master/v3/plugins/ms-vscode/vscode-github-pullrequest/0.20.0/meta.yaml'
  - type: cheEditor
    reference: https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-906/simple/che-theia-editor.yaml
```

File: dockerfiles/theia/Dockerfile [hunk 2]
```diff
@@ -9,8 +9,9 @@
 #   Red Hat, Inc. - initial API and implementation
 ###
 # Builder Image
-#
-FROM #{INCLUDE:docker/${BUILD_IMAGE_TARGET}/builder-from.dockerfile}
+#yarn
```
Comment (benoitf) [inline dockerfiles/theia/Dockerfile:12]: typo ?

Comment (vinokurig) [reply dockerfiles/theia/Dockerfile]: yes, removed the leftovers .

**Comment by benoitf** [comment]
does it require https://github.com/eclipse-theia/theia/pull/8725 ?

**Comment by vinokurig** [comment]
No, they can be merged separately

**Comment by vinokurig** [comment]
@azatsarynnyy 
> @vinokurig should the test Devfile also include Che Theia Editor from this PR? 
>```yaml
>apiVersion: 1.0.0
>metadata:
>  name: wksp-custom-dwdiu
>projects:
>  - name: che-theia
>    source:
>      location: 'https://github.com/eclipse/che-theia.git'
>      type: git
>components:
>  - type: chePlugin
>    reference: 'https://raw.githubusercontent.com/eclipse/che-plugin-registry/master/v3/plugins/ms-vscode/vscode-github->pullrequest/0.20.0/meta.yaml'
>  - type: cheEditor
>    reference: https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-906/simple/che-theia-editor.yaml
>```

Thank you for the suggestion, added it to the description.

**Review by benoitf — COMMENTED** [review]
It worked
(using the devfile of description, devfile of the latest comment has an invalid `->` string)

image

But it is asked me again to authenticate with github
image

I used the same workspace but it is asking me several times to use github authentication while I didn't restart che, etc.

**Comment by vinokurig** [comment]
The built in vscode github plugin is a separate plugin which requires another GitHub authentication session. It doesn't mean that if you signed in the GitHub Pull Requests plugin the built in GitHub plugin signed in. vscode behaves in the same way.

**Comment by benoitf** [comment]
I understand that it's on different plug-ins but still, the user experience is not good.
I'm expecting `single sign-on` between git plug-in and github PR plug-in

**Review by benoitf — APPROVED** [review]
forgot to approve this morning

**Pull request merged**